enterprise/ha/src/main/java/org/neo4j/kernel/ha/lock/forseti/ForsetiClient.java

```diff
@@ -95,96 +105,124 @@ public ForsetiClient( int id,
         }
     }
 
+    /**
+     * Reset current client state. Make it ready for next bunch of operations.
+     * Should be used before factory release client to public usage.
+     */
+    public void reset()
+    {
+        closed = false;
+        activeOperations.set( 0 );
+    }
+
     @Override
     public void acquireShared( Locks.ResourceType resourceType, long resourceId ) throws AcquireLockTimeoutException
     {
-        // Grab the global lock map we will be using
-        ConcurrentMap<Long, ForsetiLockManager.Lock> lockMap = lockMaps[resourceType.typeId()];
-
-        // And grab our local lock maps
-        PrimitiveLongIntMap heldShareLocks = sharedLockCounts[resourceType.typeId()];
-        PrimitiveLongIntMap heldExclusiveLocks = exclusiveLockCounts[resourceType.typeId()];
-
-        // First, check if we already hold this as a shared lock
-        int heldCount = heldShareLocks.get( resourceId );
-        if(heldCount != -1)
+        // client closed exiting
+        if ( closed )
         {
-            // We already have a lock on this, just increment our local reference counter.
-            heldShareLocks.put( resourceId, heldCount + 1 );
             return;
         }
-
-        // Second, check if we hold it as an exclusive lock
-        if( heldExclusiveLocks.containsKey( resourceId ) )
+        try
         {
-            // We already have an exclusive lock, so just leave that in place. When the exclusive lock is released,
-            // it will be automatically downgraded to a shared lock, since we bumped the share lock reference count.
-            heldShareLocks.put( resourceId, 1 );
-            return;
-        }
+            activeOperations.incrementAndGet();
```

Comment:
First off: This is *awesome*, I'm so happy that you're digging into the Forseti code. Please leave comments where you think it's missing as you go along as well!

However, this increment (and several others in the PR) is, unless I'm missing something, done without a guard - closed may be true at this point (eg. it was false when we checked, but before the increment above gets executed it's set to true). That means that before we increment this, we can legally be in a state where closed=true and activeOperations=0, meaning we will now be concurrently grabbing and releasing locks via the same client.

If we want to take the approach of actually allowing multi-threaded mutation of these clients, an approach that is actually both simpler and safe is to only use the atomic counter, do a CAS from `0` to `-1` to mark it as closed, replace the `closed` checks with one that the counter >= 0, and all the incrementAndGet with CAS increment loops.

*however*, I think that is a dangerous path to go down. This is one of the most complex concurrent pieces of code in the database, and if there are bugs in this code production systems will face corruption issues or full system deadlocks. Given that, I strongly feel we should apply the single-writer principle here and have it as a set-in-stone rule that these lock clients are completely owned by one thread. That would mean that rather than having the "closing" thread be allowed to directly manipulate the lock maps here, it would somehow signal to the "owning" thread that things need to shut down, and the "owning" thread would be responsible for a graceful exit. It might be enough to do those checks in the wait-loops..